src/main/java/org/jenkinsci/test/acceptance/junit/DiagnosticRule.java

```diff
@@ -0,0 +1,72 @@
+package org.jenkinsci.test.acceptance.junit;
+
+import java.io.File;
+import java.io.IOException;
+import javax.inject.Inject;
+import org.apache.commons.io.FileUtils;
+import org.jenkinsci.test.acceptance.controller.JenkinsController;
+import org.jenkinsci.test.acceptance.po.CapybaraPortingLayerImpl;
+import org.junit.rules.TestWatcher;
+import org.junit.runner.Description;
+import org.openqa.selenium.NoSuchElementException;
+import org.openqa.selenium.OutputType;
+import org.openqa.selenium.TakesScreenshot;
+import org.openqa.selenium.WebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@GlobalRule
+public class DiagnosticRule extends TestWatcher {
+    private static final Logger logger = LoggerFactory.getLogger(DiagnosticRule.class);
+
+    @Inject
+    FailureDiagnostics diagnostics;
+    @Inject
+    JenkinsController controller;
+    @Inject
+    WebDriver driver;
+
+    @Override
+    protected void failed(Throwable t, Description description) {
+        takeScreenshot();
+
+        if (causedBy(t, NoSuchElementException.class)) {
+            writeHtmlPage();
+        }
+
+        try {
+            controller.diagnose(t);
+        } catch (IOException e) {
+            throw new Error(e);
+        }
+
+    }
+
+    private void takeScreenshot() {
+        try {
+            File file = diagnostics.touch("screenshot.png");
```

Comment:
If multiple tests fail, won't the screenshots overwrite each other? This seems to be writing to the same path.

If so, you can pass in the description as an argument when invoking this method in the failed method above. Then, include the class and method name from the description, into the filename. That will avoid overwriting the same file.

Reply:
the touch() method is putting the file in a directory with the name of the test, as we take only one screenshot by failing test it should be ok.

Reply:
@romainverduci okay perfect